### PR TITLE
Add runs_before Builder config and use it for ordering

### PR DESCRIPTION
**Pull request by natebosch**
Towards #1065

- Add `runs_before` parsing and include it in the test.
- Don't split builders into groups based on whether they have
  `requires_input`. This could slightly change ordering of some builders
  but not in a way that has any negative impact.
- Use `runs_before` to set dependencies for builders when ordering them.

### CODE REVIEW

**Comment by natebosch** [comment]
cc @nshahan

**Comment by nshahan** [comment]
You beat me to the PR. :smiley:

I am still having a confusing issue with a file not being ready for the angular or ddc builders at the right time in the gallery project I am working on. Same thing with your version and mine. Maybe you can look at it with me tomorrow.

File: build_runner/test/build_script_generate/builder_ordering_test.dart [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+import 'package:test/test.dart';
+
+import 'package:build_runner/src/build_script_generate/builder_ordering.dart';
+
+import '../common/build_configs.dart';
+
+void main() {
+  group('Builder ordering', () {
+    test('orders builders with `runs_before`', () async {
```
Comment (jakemac53) [inline build_runner/test/build_script_generate/builder_ordering_test.dart]: nit: add some error validations tests for a cycle created by `runs_before` and `required_inputs`?

**Comment by nshahan** [comment]
Deleting the .dart_tool directory is what finally got the missing file to generate for me. I'm not sure if that was overkill but now I can see the order of the builders get changed as I would expect it to with this change.

File: build_runner/test/build_script_generate/builder_ordering_test.dart (hunk 1 above)
Comment (nshahan) [reply build_runner/test/build_script_generate/builder_ordering_test.dart]: nit: add a warning to the documentation that describes a the danger of creating a cycle.

Comment (natebosch) [reply build_runner/test/build_script_generate/builder_ordering_test.dart]: Added a separate section in the build_config README about builder ordering. Added a test that when there is a cycle it throws.

**Pull request merged**